### ISSUE TITLE
Streaming Profiler

### ISSUE THREAD

**Issue by nickwanninger** [SerenityOS/serenity#2673]
I haven't really fully investigated the kernel-side of the profiler, but would it be possible for the profiler to consume the kernel's stacktrace buffer when you read from the `/proc` entry? This would allow for the Profiler application to "stream" new profiler information occasionally from the `/proc/` file and allow the profiler to run for longer.

The process being profiled could also block until there is space in the stacktrace buffer so you don't miss out on any stack frames.

**Comment by alimpfard** [comment]
This is dependent on two things (that I can think of rn)
- ability to stream json
- ability to parse streaming json

while I _do_ have a (rather dated) stream json parser, we don't have a way to actually stream the events yet.
I believe @bugaevc worked on streaming data out at some point? (summoning for clarification)

other than that, I don't think there's anything _special_ that has to be done about this, `/proc/profile` can just be read in nonblock mode in an event loop if blocking isn't desired (and I don't think it is).

**Comment by nickwanninger** [comment]
I dont think it needs to "stream" the json. It could just be that when you read from /proc/profile it consumes the profile information (clearing it from the kernel buffer)

**Comment by alimpfard** [comment]
something something, anything worth doing is worth overdoing?

but I see what you mean, and...isn't it just how it is already?
I'm not sure if the stuff in `/proc/profile` persist after being read

**Comment by nickwanninger** [comment]
procfs$profile() doesn't clear the backtrace buffer. I am picturing the profiler updating it's information as it is profiling, displaying new information every 100ms or something until you ask it to stop profiling. This could last forever until the userspace process runs out of memory because the profiler in the kernel would be allowed to clear out it's `s_profiling_buffer` whenever the Profiler reads it.

**Comment by bugaevc** [comment]
> while I do have a (rather dated) stream json parser, we don't have a way to actually stream the events yet.
> I believe @bugaevc worked on streaming data out at some point? (summoning for clarification)

Yeah 🙂 so we do have kernel-side streaming stuff, but it stalled on streaming JSON parser because I couldn't come up with a good enough API for it. https://github.com/SerenityOS/serenity/pull/1167

**Comment by alimpfard** [comment]
We tend to disagree on APIs, but here's the one I had come up with (as I remember it, been a while):
```c++
// look at every <root>.foo.bar[x], whatever index x may be
// everything else would be saved by default
parser.stream(JsonPath {"foo", "bar", JsonPathElement::any_index}, [](auto object) { 
    // do whatever you want with the object
   return VisitDecision::Discard;
});
parser.somehow_be_fed_data();
```

**Comment by bugaevc** [comment]
> We tend to disagree on APIs

That's a good thing! We have a saying here (apparently, Socrates said it first?) that the truth is born in dispute/argument :smile:

I've considered these two styles:

```cpp
parser.on_root_object = [](auto& object_parser) {
     object_parser.on_key_value_pair = [](auto& key, auto& value_parser) {
         // ...
     };
};

block_waiting_for_data();
parser.heres_some_data_for_you();
block_waiting_for_moar_data();
parser.heres_some_moar_data_for_ya();
```

and

```cpp
parser.on_need_more_data = [] {
    block_waiting_for_moar_data();
};

for (auto& key_value_pair : parser.root_object()) {
    // ...
}
```

The latter is almost as convenient as the "upfront" JSON parsers we have today, but I imagine it's almost impossible to integrate into a EventLoop-based app like Profiler. The former is clearly more flexible, but it requires tons of nested lambdas, and it's not at all convenient to use (to e.g. iterate over a properties of an object).

**Comment by alimpfard** [comment]
Both of those...I hate.
you're streaming only the immediate children of the root node, which is not always desired.

I personally prefer telling the parser upfront "feed the value of this path into this function", which would allow for fun stuff like "gimme only `foo.bar.baz.<x>.whatever[y]`", and you don't even have to worry about all the other stuff.

--

Although this doesn't really matter for the Profile~~Viewe~~r
https://github.com/alimpfard/serenity/blob/c00a9a9ee1494bbdff2a2f6f20cb05a2711af137/DevTools/ProfileViewer/Profile.cpp#L199-L268

**Comment by bugaevc** [comment]
> you're streaming only the immediate children of the root node, which is not always desired.

No, I've only shown the root node for the example, but it of course should work uniformly for any object/array anywhere in the hierarchy. That's why I'm saying the former approach looks nightmarish to iterate over a moderately complex data structure.

> I personally prefer telling the parser upfront "feed the value of this path into this function", which would allow for fun stuff like "gimme only foo.bar.baz.<x>.whatever[y]", and you don't even have to worry about all the other stuff.

But that's not how we use JSON parsing in practice. We iterate over an array, create some sort of structure for each element, then fill it with data from an element (which would be an object), then perhaps parse some child data in a similar way. See https://github.com/SerenityOS/serenity/blob/master/Libraries/LibCore/ProcessStatisticsReader.cpp for a typical example.

**Comment by alimpfard** [comment]
> it should work uniformly for any object/array anywhere in the hierarchy

Not sure how it would look, but sure.

> the former approach looks nightmarish

Agreed on that :)

> But that's not how we use JSON parsing in practice [...] See [there] for a typical example

but that _is_ how we use json in practice, we rarely if ever care about the siblings of something in the hierarchy.
e.g. the user of `[1]` has no interest in `[0]`, but all the interest in `[1].pid` and `[1].threads[2]`.
and if you ask for `[1]`, you get _all_ of its subtree, but nothing outside that.

the API you're proposing is nice for a low-level stream json parser (see https://github.com/SerenityOS/serenity/pull/2019/commits/09b34f508b7ece18f6c3d5458ef059e22e52f84f#diff-a6f02ea9b1a60e7f5401682ebcf42412R51-R58 for the low-level API I impl'd, it's similar to your first method)
but really, any way we can think of that leaks the json structure out (e.g. `on_key_value`) would be hell to work with normally.

**Comment by bugaevc** [comment]
> we rarely if ever care about the siblings of something in the hierarchy. e.g. the user of [1] has no interest in [0], but all the interest in [1].pid and [1].threads[2].

I guess, but I don't see how's that important?

As for your API plan, I don't understand what's the point of the "JSON paths". We almost never want just some subset of the data; we want to parse each item, fully, as it comes, right?

**Comment by alimpfard** [comment]
We do parse everything as it comes, but we often have specific uses for each specific path in the hierarchy, and we often just sequentially read them, using the paths makes this a bit more bearable, something like `on_executable = ...` in  a ProfileReader context 
I find it significantly more ergonomic than "on_key_value = [](key, value) { if (key == "executable") {...})`

**Comment by nickwanninger** [comment]
My idea wasn't to actually stream (my bad on the issue title) json from the kernel, but to have seperate read calls to /proc/profile return the entire state of the profiler each time, causing the profiler to be emptied so the kernel can gather more stackframe data. The problem is that the kernel has a finite size buffer for the profiler information but we might want to analyze a larger timeslice

**Comment by alimpfard** [comment]
I understood that, and I'm pretty sure Sergey did too, this was just our typical _argue over random things in random places_ thing that often ends up causing either of us to implement the thing in question.

Your idea is a pretty good in-between though.

**Comment by nickwanninger** [comment]
ah okay :^)

**Comment by bugaevc** [comment]
Yeah, I sure do understand that your proposal wasn't about actually streaming; but given how I have already implemented streaming profiler kernel-side, and Ali has already written a streaming parser (even if we haven't reached a consensus on API yet), it makes sense for us to aim directly for actual streaming.
